enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,402 @@
+- The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85) to share a common approach with the bootstrap specific static-pod configuration.
+- Operator will be built with a goal to achieve provider decoupling and potential switching between providers in the cluster.
+
+**TBD: Cluster-credentials-operator integration to fetch cloud credentials**
+
+## Upgrade/Downgrade strategy
+
+A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.
+
+Upgrade from previous versions of OpenShift on OpenStack will look like:
+
+- On the initial stage of upgrading `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
+
+- `cluster-cloud-controller-manager-operator` creates all required resources for CCM (RBAC, Service Account, etc.).
+
+- `cluster-cloud-controller-manager-operator` creates `deployments` for CCM.
+
+- `kube-apiserver` and `kube-controller-manager` are restarted without the `--cloud-provider` option.
+
+- `kubelet` is restarted with `--cloud-provider=external` option.
+
+Downgrade:
+
+- An old machine configuration is applied, which causes `kubelet` to restart with `--cloud-provider=some-platform` option.
+
+- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider=some-platform` option.
+
+- `cluster-cloud-controller-manager-operator` stops working.
+
+### Version Skew Strategy
+
+See the upgrade/downgrade strategy.
+
+### Action plan (for OpenStack)
+
+#### Build OpenStack CCM image by OpenShift automation
+
+To start using OpenStack CCM in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The CCM image should be automatically tested before it becomes available.
+The upstream repo provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in [Dockerhub](https://hub.docker.com/r/k8scloudprovider/openstack-cloud-controller-manager).
+
+Actions:
+
+- Configure CI operator to build OpenStack CCM image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OpenShift Quay account.
+
+#### Test Cloud Controller Manager manually
+
+When all required components are built, we can manually deploy OpenStack CCM and test how it works.
+
+Actions:
+
+- Manually install CCM's [daemonset](https://github.com/kubernetes/cloud-provider-openstack/blob/master/manifests/controller-manager/openstack-cloud-controller-manager-ds.yaml) on a working OpenShift cluster deployed on OpenStack.
```

Comment:
Should we update this to a deployment to more closely mimic what the operator would do?

Reply:
They use a `daemonSet` for that example, and I assume will yet require some changes for the config maps mounts used in that pod configuration, as well as other arguments. Do you think this section should be included in the proposal, @Fedosin?